// 实现一个命令的接口类

syntax = "proto3";

package command;

message Command
{
    
    // uint32 version = 1;  // raft 节点各自独立使用命令，不存在并发控制，不需要使用乐观锁
    oneof command_type
    {
        PutCommand put = 1;
        GetCommand get = 2;
        DeleteCommand del = 3; // delete 是 proto 的关键字，不能用作标识符
        IncrementCommand increment = 4;
    }
}

message PutCommand
{
    bytes key = 1;
    bytes value = 2;
}

message GetCommand
{
    bytes key = 1;
}

message DeleteCommand
{
    bytes key = 1;
    uint32 version = 2;
}

message IncrementCommand
{
    bytes key = 1;
    int64 delta = 2;
}